storybookjs/storybook
# Theming: Add theme variable to set the preview background color

Closes #22029 






## What I did

This PR introduces a new theming variable, `appPreviewBg` to explicitly set the preview background color. This was previously possible with `appContentBg` in 6.5, but that was problematic as documented in https://github.com/storybookjs/storybook/issues/21080.

We generally recommend setting explicit backgrounds in the rendered stories directly (eg. with a decorator in `preview.ts`), but doing this directly in the manager is now also possible with `appPreviewBg`.

**an open question is whether or not the "loading preview" element should have the same preview background or maintain the current app background. in dark mode, the current app background is dark so the loading screen will be dark for a short period of time before going white. if we changed it to match, it would be white all the time.**



## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

See this demonstrated in different scenarios in this sandbox: https://stackblitz.com/edit/github-u7boko?file=.storybook%2Fpreview.ts

1. `@storybook/addon-themes` has been added to demonstrate that any background set using that addon will still override whatever background set in the manager with the theme variable. Story/preview level styles should always override manager styles. You need to disable this addon to experience the next two scenarios.
2. `storybook-dark-mode` has been added to demonstrate that it still works to set manager backgrounds from addons, if the user specifies the new theme variable in the parameters (as documented by the addon). You need to disable this addon to experience the next scenario
3. Users can now set the preview background directly with the usual `addons.setConfig` API. (Changing this requires restarting the SB server)
4. Finally disable all of the above to see that previews still default to a white background, both when Storybook is in dark and light mode - use the browser DevTools to toggle between the two and reload the site).



### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release


This pull request has been released as version [`0.0.0-pr-24575-sha-94eee1ff`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24575-sha-94eee1ff). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24575-sha-94eee1ff`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24575-sha-94eee1ff) |
| **Triggered by** | @JReinhold |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`22029-bug-preview-always-has-a-white-background`](https://github.com/storybookjs/storybook/tree/22029-bug-preview-always-has-a-white-background) |
| **Commit** | [`94eee1ff`](https://github.com/storybookjs/storybook/commit/94eee1ffe56b1cd3c993e1ed50c9620930ec551b) |
| **Datetime** | Wed Oct 25 21:46:11 UTC 2023 (`1698270371`) |
| **Workflow run** | [6646514771](https://github.com/storybookjs/storybook/actions/runs/6646514771) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24575`_
</details>
